src/core/public/osd_bootstrap.ts

```diff
@@ -31,13 +31,41 @@
 import { i18n } from '@osd/i18n';
 import { CoreSystem } from './core_system';
 import { ApmSystem } from './apm_system';
+import { getAndUpdateLocaleInUrl } from './locale_helper';
 
 /** @internal */
 export async function __osdBootstrap__() {
   const injectedMetadata = JSON.parse(
     document.querySelector('osd-injected-metadata')!.getAttribute('data')!
   );
 
+  // Extract the locale from the URL if present
+  // This allows for dynamic locale setting via URL parameters
+  const urlLocale = getAndUpdateLocaleInUrl(window.location.href);
+
+  if (urlLocale) {
+    // If a locale is specified in the URL, update the i18n settings
+    // This enables dynamic language switching
+    // Note: This works in conjunction with server-side changes:
+    // 1. The server registers all available translation files at startup
+    // 2. A server route handles requests for specific locale translations
+
+    // Set the locale in the i18n core
+    // This will affect all subsequent i18n.translate() calls
+    i18n.setLocale(urlLocale);
+
+    // Modify the translationsUrl to include the new locale
+    // This ensures that the correct translation file is requested from the server
+    // The replace function changes the locale in the URL, e.g.,
+    // from '/translations/en.json' to '/translations/zh-CN.json'
+    injectedMetadata.i18n.translationsUrl = injectedMetadata.i18n.translationsUrl.replace(
+      /\/([^/]+)\.json$/,
+      `/${urlLocale}.json`
+    );
+  } else {
+    i18n.setLocale('en');
```

Comment:
why is this needed?

Reply:
I remember this is for user removing the `i18n-locale` part from url after first add it.